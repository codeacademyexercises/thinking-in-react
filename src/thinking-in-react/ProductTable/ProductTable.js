import React ,{ Component } from 'react';
import ProductCategoryRow from '../ProductCategoryRow/ProductCategoryRow.js';

class ProductTable extends Component{
    store = [{category: 'Sporting Goods',Name: 'Football',Price: 49.99, stocked: true},
    {category: 'Sporting Goods',Name: 'Baseball',Price: 9.99, stocked: true},
    {category: 'Sporting Goods',Name: 'Basketball',Price: 29.99, stocked: false},
    {category: 'Electronics',Name: 'iPod Touch',Price: 99.99, stocked: true},
    {category: 'Electronics',Name: 'iPhone 5',Price: 399.99, stocked: false},
    {category: 'Electronics',Name: 'Nexus 7',Price: 199.99, stocked: true},
    ];
    InterestedItems = (regex,instock)=> this.store.reduce((acc, item)=>{
        if(item.category.search(regex)!==-1 || item.Name.search(regex)!==-1){
            if(instock===true){
                if(item.stocked===true){
                    acc.push(item);
                }
            }else{
            acc.push(item);
            }
        }
        return acc;
    },[]);
    Categories = (result) => {
        return result.reduce((acc,item)=>{
            if(acc.indexOf(item.category)===-1){
                acc.push(item.category);
            }
            return acc;
        },[]);
    };
    render(){
        console.log(this.props.search,this.props.stock);
        let display;
        let regex = new RegExp("^"+this.props.search, "i");
        let instock = this.props.stock;
        let result = this.InterestedItems(regex,instock);
        let categories = this.Categories(result);
        if(result.length===0){
            display="Srry No Items Found...";
        }else{
            display = [];
            categories.forEach((category)=>{
                display.push(<ProductCategoryRow key={category} category={category} product={result}/>);
            });
            display.splice(0,<table>);
            display.splice(1,<tr key="ProductHeader">
                <th>Name</th><th>Price</th></tr>);
                display.splice(1,<tbody>);
                    display.push(</tbody>)
                display.push(</table>);
        }
        return(
            <div className="tableDetails">
                {display}
            </div>
        );
    }
};

export default ProductTable;