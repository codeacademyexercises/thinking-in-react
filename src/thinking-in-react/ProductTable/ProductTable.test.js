import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import ProductTable from './ProductTable.js';

describe("<ProductTable /> should ",()=>{
    it("render",()=>{
        const tree = renderer.create(<ProductTable search="" stock={true}/>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('have functions working correctly',()=>{
        const wrapper = shallow(<ProductTable/>);
        wrapper.setProps({
            search: "Electronics",
            stock: true,
        });
        expect(wrapper.find('ProductCategoryRow').props().product).toHaveLength(2);
        expect(wrapper.find('ProductCategoryRow').props().category).toEqual("Electronics");
        wrapper.setProps({
            search: "z",
            stock: true,
        });
        expect(wrapper.text()).toEqual("Srry No Items Found...");
    });
});