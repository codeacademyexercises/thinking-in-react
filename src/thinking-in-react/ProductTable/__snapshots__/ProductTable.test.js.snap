// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTable /> should  render 1`] = `
<div
  className="tableDetails"
>
  <div>
    <table>
      <tbody>
        <tr>
          <th>
            Sporting Goods
          </th>
        </tr>
        <tr>
          <td>
            Football
          </td>
          <td>
            $
            49.99
          </td>
        </tr>
        <tr>
          <td>
            Baseball
          </td>
          <td>
            $
            9.99
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div>
    <table>
      <tbody>
        <tr>
          <th>
            Electronics
          </th>
        </tr>
        <tr>
          <td>
            iPod Touch
          </td>
          <td>
            $
            99.99
          </td>
        </tr>
        <tr>
          <td>
            Nexus 7
          </td>
          <td>
            $
            199.99
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
